src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,625 @@
+        if (!data)
+                return log_oom();
+
+        pos = data;
+        memcpy(pos, &w->dev, sizeof(struct wgdevice));
+        pos = pos + sizeof(struct wgdevice);
+        LIST_FOREACH(peers, peer, w->peers) {
+                memcpy(pos, &peer->fields, sizeof(struct wgpeer));
+                pos = pos + sizeof(struct wgpeer);
+
+                LIST_FOREACH(ipmasks, mask, peer->ipmasks) {
+                        memcpy(pos, &mask->fields, sizeof(struct wgipmask)),
+                        pos = pos + sizeof(struct wgipmask);
+                }
+        }
+        assert((size_t)((char*)pos - (char*)data) == w->allocation_size);
+        ifreq.ifr_data = data;
+
+        if (ioctl(fd, WG_SET_DEVICE, &ifreq) == -1)
+                return log_netdev_error_errno(netdev, errno, "Unable configure wireguard device: %m");
+
+        return 0;
+}
+
+static void endpoint_free(WireguardEndpoint *e) {
+        assert(e);
+        free(e->host);
+        free(e->port);
+        free(e);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(WireguardEndpoint*, endpoint_free);
+
+static int on_resolve_retry(sd_event_source *s, usec_t usec, void *userdata) {
+        NetDev *netdev = userdata;
+        Wireguard *w;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        w->resolve_retry_event_source = sd_event_source_unref(w->resolve_retry_event_source);
+
+        w->unresolved_endpoints = w->failed_endpoints;
+        w->n_unresolved_endpoints = w->n_failed_endpoints;
+        w->failed_endpoints = NULL;
+        w->n_failed_endpoints = 0;
+
+        resolve_endpoints(netdev);
+
+        return 0;
+}
+
+static int wireguard_resolve_handler(sd_resolve_query *q,
+                                     int ret,
+                                     const struct addrinfo *ai,
+                                     void *userdata) {
+        NetDev *netdev;
+        Wireguard *w;
+        _cleanup_(endpoint_freep) WireguardEndpoint *e;
+        int r;
+
+        assert(userdata);
+        e = userdata;
+        netdev = e->netdev;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_REMOVE(endpoints, w->unresolved_endpoints, e);
+        w->n_unresolved_endpoints--;
+
+        if (ret != 0) {
+                log_netdev_error(netdev, "Failed to resolve host '%s:%s': %s", e->host, e->port, gai_strerror(ret));
+                LIST_PREPEND(endpoints, w->failed_endpoints, e);
+                w->n_failed_endpoints++;
+                e = NULL;
+        } else if ((ai->ai_family == AF_INET && ai->ai_addrlen == sizeof(struct sockaddr_in)) ||
+                        (ai->ai_family == AF_INET6 && ai->ai_addrlen == sizeof(struct sockaddr_in6)))
+                memcpy(&e->peer->fields.endpoint, ai->ai_addr, ai->ai_addrlen);
+        else
+                log_netdev_error(netdev, "Neither IPv4 nor IPv6 address found for peer endpoint: %s:%s", e->host, e->port);
```

Comment:
This error message can be quite misleading, when the address is there but for whatever reason the size does not match. I think it'd be better to make the error messages match the actual conditions more closely.
```C
if (ret != 0) {
    ...
} else if (IN_SET(ai->ai_family, AF_INET, AF_INET6)) {
         if (ai->ai_addrlen != FAMILY_ADDRESS_SIZE(ai->family))
                 log_netdev_error(netdev, "Wrong address size for family %s, (%zu, expected %zu)", ...);
          memcpy(...);
} else
      log_netdev_error(netdev, "Address from unsupported family %s found for peer endpoint: %s:%s", ...);
```

Another question: shouldn't this function return immediately on error here?